keps/sig-network/3015-prefer-same-node/README.md

```diff
@@ -0,0 +1,496 @@
+type EndpointHints struct {
+        ...
+
+	// forNodes indicates the node(s) this endpoint should be targeted by.
+	// +listType=atomic
+	ForNodes []string `json:"forNodes,omitempty" protobuf:"bytes,2,name=forNodes"`
+}
+```
+
+When updating EndpointSlices, if the EndpointSlice controller sees a
+service with `PreferSameNode` traffic distribution, then for each
+endpoint in the slice, it will add a `ForNodes` hint including the
+name of the endpoint's node. (The field is an array for future
+extensibility, but initially it will always have either 0 or 1
+elements.) In addition, it will set the `ForZones` hint as it would
+with `TrafficDistribution: PreferClose`, to allow older service
+proxies to fall back to at least same-zone behavior.
+
+Kube-proxy's `CategorizeEndpoints` function will be updated as
+follows:
+
+  - If every endpoint has a `ForNodes` hint set, and at least one
+    `ForNodes` hint includes the local node, then the set of
+    "topologically available" endpoints is the set of all endpoints
+    with a `ForNodes` hint that includes the local node.
+
+  - Otherwise, if every endpoint has a `ForZones` hint set, and at
+    least one `ForZones` hint includes the local node's zone, then the
+    set of "topologically available" endpoints is the set of all
+    endpoints with a `ForZones` hint that includes the local node's
+    zone.
+
+  - Otherwise, all endpoints are "topologically available".
+
+(The first step is new; the other two steps reflect the current code.)
+
+### Test Plan
+
+[X] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+N/A
+
+##### Unit tests
+
+Tests of validation, endpointslice-controller, and kube-proxy will be
+updated.
+
+<!--
+Additionally, for Alpha try to enumerate the core package you will be touching
+to implement this enhancement and provide the current unit coverage for those
+in the form of:
+- <package>: <date> - <current test coverage>
+The data can be easily read from:
+https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit
+
+This can inform certain test coverage improvements that we want to do before
+extending the production code to implement this enhancement.
+-->
+
+- `<package>`: `<date>` - `<test coverage>`
+
+##### Integration tests
+
+N/A
+
+##### e2e tests
+
+E2E tests will be added similar to existing traffic distribution
+tests, to cover the new options.
+
+- <test>: <link to test coverage>
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+
+- Unit tests for API enablement and endpoint selection.
+
+#### Beta
+
+- E2E tests completed and enabled.
+
+#### GA
+
+- Time passes, no major objections
+
+### Upgrade / Downgrade Strategy
+
+No real issues, other than dealing with version skew.
```

Comment:
I agree it's all standard stuff, but technically you're extending the API (the hints struct), so please elaborate a bit more about it.